ghcide/src/Development/IDE/Core/Compile.hs

```diff
@@ -179,16 +187,102 @@ typecheckModule (IdeDefer defer) hsc keep_lbls pm = do
     where
         demoteIfDefer = if defer then demoteTypeErrorsToWarnings else id
 
--- | Add a Hook to the DynFlags which captures and returns the
--- typechecked splices before they are run. This information
--- is used for hover.
-captureSplices :: HscEnv -> (HscEnv -> IO a) -> IO (a, Splices)
-captureSplices env k = do
+-- | Install hooks to capture the splices as well as the runtime module dependencies
+captureSplicesAndDeps :: HscEnv -> (HscEnv -> IO a) -> IO (a, Splices, UniqSet ModuleName)
+captureSplicesAndDeps env k = do
   splice_ref <- newIORef mempty
-  res <- k (hscSetHooks (addSpliceHook splice_ref (hsc_hooks env)) env)
+  dep_ref <- newIORef emptyUniqSet
+  res <- k (hscSetHooks (addSpliceHook splice_ref . addLinkableDepHook dep_ref $ hsc_hooks env) env)
   splices <- readIORef splice_ref
-  return (res, splices)
+  needed_mods <- readIORef dep_ref
+  return (res, splices, needed_mods)
   where
+    addLinkableDepHook :: IORef (UniqSet ModuleName) -> Hooks -> Hooks
+    addLinkableDepHook var h = h { hscCompileCoreExprHook = Just (compile_bco_hook var) }
+
+    -- We want to record exactly which linkables/modules the typechecker needed at runtime
+    -- This is useful for recompilation checking.
+    -- From hscCompileCoreExpr'
+    -- Only compute direct dependencies instead of transitive dependencies.
+    -- It is much cheaper to store the direct dependencies, we can compute
+    -- the transitive ones when required.
+    -- Also only record dependencies from the home package
+    compile_bco_hook :: IORef (UniqSet ModuleName) -> HscEnv -> SrcSpan -> CoreExpr -> IO ForeignHValue
+    compile_bco_hook var hsc_env srcspan ds_expr
```

Comment:
I imagine that the body of this function is lifted from the default implementation of this hook in GHC.

Could you add a reference together with instructions on how to update it on subsequent GHC versions?